src/metabase/api/collection.clj

```diff
@@ -296,16 +298,38 @@
     (not (zero? v))
     v))
 
-(defn- has-required-parameters? [row]
-  (if-let [template-tags (-> row :dataset_query (json/parse-string keyword) :native :template-tags)]
-    (every? #(or (not (:required %)) (:default %)) (vals template-tags))
-    true))
+(defn- fully-parametrized?
+  "Decide if the query in the card represented by the result row `row` is fully parametrized.
+
+  The rules to consider a query fully parametrized is as follows:
+
+  1. All parameters not in an optional block are field-filters or have a default value.
+  2. All required parameters have a default value.
+
+  The first rule is absolutely necessary, as queries violating it cannot be executed without
+  externally supplied parameter values. The second rule is more controversial, as field-filters
+  outside of optional blocks ([[ ... ]]) don't prevent the query from being executed without
+  external parameter values (neither do parameters in optional blocks). The rule has been added
+  nonetheless, because marking a parameter as required is something the user does intentionally
+  and queries that are technically executable without parameters can be unacceptably slow
+  without the necessary constraints. (Marking parameters in optional blocks as required doesn't
+  seem to be useful any way, but if the user said it is required, we honor this flag.)"
+  [row]
+  (let [native-query (-> row :dataset_query (json/parse-string keyword) :native)]
+    (if-let [template-tags (:template-tags native-query)]
+      (let [obligatory-params (into #{}
+                                    (comp (filter params/Param?)
+                                          (map (comp keyword :k)))
+                                    (-> native-query :query params.parse/parse))]
```

Comment:
Actually, @qnkhuat is currently working on allowing parameters to be defined in snippets in https://github.com/metabase/metabase/pull/23658. From that PR it seems like this might not work automatically for those cases.

Not sure if there's an easy way to reconcile these two PRs, other than adding more code here to recursively parse params from snippets so that they can be considered for the `fully_parameterized` value.

What do you think @qnkhuat?

Reply:
I've seen that PR too. I wonder if we could have a function which "flattens" the query by resolving the snippets without information about supplied parameters.

Reply:
Merging this with the note that queries with snippets will never be considered fully parametrized.

Reply:
Sorry for the late reply, but I think we could definitely do that without information about parameters.
Basically walk through the template tags, if it's a snippet then fetch the snippet, do another parse and walk through the snippet's snippets content.

But I believe that is not the scope of this PR just yet. I'll see if we need to add it once the parameters in snippet is merged.

Also note that we can do the same for `cardReference` parameters.

> Not sure if there's an easy way to reconcile these two PRs

I'm afraid we can't since my PR needs to have the `parameters` to construct a `tag->value` map.